glib/src/subclass/shared.rs

```diff
@@ -0,0 +1,227 @@
+where
+    T: 'static,
+{
+    type InnerType = T;
+
+    fn ref_(this: *const Self::InnerType) -> *const Self::InnerType {
+        unsafe {
+            use std::mem::ManuallyDrop;
+            let this_rc = ManuallyDrop::new(std::rc::Rc::from_raw(this));
+            std::rc::Rc::into_raw(ManuallyDrop::take(&mut this_rc.clone()))
+        }
+    }
+
+    fn into_ptr(self) -> *const Self::InnerType {
+        std::rc::Rc::into_raw(self)
+    }
+
+    fn from_ptr(this: *const Self::InnerType) -> Self {
+        unsafe { std::rc::Rc::from_raw(this) }
+    }
+}
+
+/// Trait for defining shared types.
+///
+/// Links together the type name with the type itself.
+///
+/// See [`register_shared_type`] for registering an implementation of this trait
+/// with the type system.
+///
+/// [`register_shared_type`]: fn.register_shared_type.html
+pub trait SharedType: Clone + Sized + 'static {
+    /// Shared type name.
+    ///
+    /// This must be unique in the whole process.
+    const NAME: &'static str;
+
+    /// The inner refcounted type
+    type RefCountedType: RefCounted;
+
+    /// Returns the type ID.
+    ///
+    /// This is usually defined via the [`Shared!`] derive macro.
+    ///
+    /// [`Shared!`]: ../../derive.Shared.html
+    fn get_type() -> crate::Type;
+
+    /// Converts the SharedType into its inner RefCountedType
+    fn into_refcounted(self) -> Self::RefCountedType;
+
+    /// Constructs a SharedType from a RefCountedType
+    fn from_refcounted(this: Self::RefCountedType) -> Self;
+}
+
+/// Register a boxed `glib::Type` ID for `T`.
+///
+/// This must be called only once and will panic on a second call.
+///
+/// See [`Shared!`] for defining a function that ensures that
+/// this is only called once and returns the type id.
+///
+/// [`Shared!`]: ../../derive.Shared.html
+pub fn register_shared_type<T: SharedType>() -> crate::Type {
+    unsafe {
+        use std::ffi::CString;
+        unsafe extern "C" fn shared_ref<T: SharedType>(v: ffi::gpointer) -> ffi::gpointer {
+            T::RefCountedType::ref_(v as *const <T::RefCountedType as RefCounted>::InnerType)
+                as ffi::gpointer
+        }
+        unsafe extern "C" fn shared_unref<T: SharedType>(v: ffi::gpointer) {
+            let _ = T::RefCountedType::from_ptr(
+                v as *const <T::RefCountedType as RefCounted>::InnerType,
+            );
+        }
+
+        let type_name = CString::new(T::NAME).unwrap();
+        if gobject_ffi::g_type_from_name(type_name.as_ptr()) != gobject_ffi::G_TYPE_INVALID {
+            panic!(
+                "Type {} has already been registered",
+                type_name.to_str().unwrap()
+            );
+        }
+
+        from_glib(gobject_ffi::g_boxed_type_register_static(
+            type_name.as_ptr(),
+            Some(shared_ref::<T>),
+            Some(shared_unref::<T>),
+        ))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    // Shared macro assumes 'glib' is in scope
```

Comment:
It really shouldn't do that :) Look at the `#crate_ident` hack used in the other macros to avoid that.

Reply:
While at it, please fix the same for the `GBoxed` derive too

Reply:
I looked into that, but it seems [it hardcodes the `crate_ident` to `glib` in case the macro is being invoked inside the `glib` crate](https://github.com/gtk-rs/gtk-rs/blob/master/glib-macros/src/utils.rs#L155-L157). I guess it shoud be hardcoded to `crate` instead - I tried to change it but got a bunch of other errors in the `clone` macro:
```
error[E0433]: failed to resolve: maybe a missing crate `g_debug`?
  --> glib/tests/clone.rs:31:9
   |
31 | /         clone!(@weak ref_state => move || {
32 | |             ref_state.borrow_mut().started = true;
33 | |         })
   | |__________^ maybe a missing crate `g_debug`?

```
so I left it like that for the moment.
Btw [also the `GBoxed` macro does a similar thing currently](https://github.com/gtk-rs/gtk-rs/blob/master/glib/src/subclass/boxed.rs#L70-L71)

Reply:
That should be indeed `crate` in that case. For the `clone!` macro, can you give the `cargo expand` code of that part?

Reply:
Sure, here it is.
Looking deeper into it, all the integration tests in the `tests` subfolder of `glib` are failing, because one cannot use the `crate::` prefix in integration tests to refer to the crate being tested.

[cargo_expand.txt](https://github.com/gtk-rs/gtk-rs/files/6317144/cargo_expand.txt)

Reply:
But in the integration tests the crate is available with its own name (`glib`). So I assume our problem is that in the unit tests with need `crate` and in the integration tests we need `glib`, and we can't distinguish which of the two cases it is from the macro?

In that case I would suggest to keep it returning `glib` and add a short explanatory comment into the unit tests about this. Still use the `crate_ident` magic inside the macro so it works outside of glib properly :)

Reply:
Yes that's the problem. Sure, I will add that comment and double-check that I'm using `crate_ident` inside the macro everywhere :)

Reply:
Great, thanks :)

Reply:
Comments added, and macro checked - should be ok now